src/s_inter.c

```diff
@@ -138,6 +203,514 @@ void sys_set_temppath(void);
+void sys_addsendfdrmfn(int fd, t_fdsendrmfn fn, void* ptr)
+{
+    int nfd, size;
+    t_fdsend *fs;
+    if (!pd_this->pd_inter->i_fdsend)
+    {
+        /* create an empty FD list */
+        pd_this->pd_inter->i_fdsend = (t_fdsend *)t_getbytes(0);
+        pd_this->pd_inter->i_nfdsend= 0;
+    }
+    if(rbrmfdsend_init())
+        return;
+    nfd = pd_this->pd_inter->i_nfdsend;
+    size = nfd * sizeof(t_fdsend);
+    pd_this->pd_inter->i_fdsend = (t_fdsend *)t_resizebytes(
+        pd_this->pd_inter->i_fdsend, size, size + sizeof(t_fdsend));
+    fs = pd_this->pd_inter->i_fdsend + nfd;
+    fs->fds_fd = fd;
+    fs->fds_fn = fn;
+    fs->fds_ptr = ptr;
+    pd_this->pd_inter->i_nfdsend = nfd + 1;
+}
+
+// a partial replacement for the functionalities of sys_domicrosleep() when
+// enabling THREADED_IO. Namely:
+// - check for ready file descriptors, and for each ready one:
+//  - if it has a RT-friendly callback, retrieve the new data from the
+//  socket and place it in a ring buffer. sys_domicrosleep() will then call the
+//  callback which will retrieve the data from the buffer
+//  - otherwise, mark it so that the callback will be called from
+//  sys_domicrosleep()) at the first occasion
+//
+// As a consequence of this:
+//  - for callbacks that are RT-friendly, sys_domicrosleep() does not need to select()
+//  - for callbacks that are not RT-friendly, sys_domicrosleep() doesn't have to select()
+//  at every call, but only when data is actually available.
+static void poll_fds()
+{
+    const unsigned int bufsize = pd_this->pd_inter->i_iothreadbufsize;
+    ssize_t ret;
+    struct timeval timout = {0}; // making this timeout longer would be more efficient (by avoiding spurious wakeups), but it would needlessly postpone writes
+    int pollem = 1;
+    int i;
+    t_fdpoll *fp;
+    if (pollem && pd_this->pd_inter->i_nfdpoll)
+    {
+        fd_set readset;
+        fdp_select(&readset, timout, kFdpManagerIoThread);
+        for (i = 0; i < pd_this->pd_inter->i_nfdpoll; i++)
+        {
+            t_fdpoll *fp = pd_this->pd_inter->i_fdpoll + i;
+            int fd = fp->fdp_fd;
+            if(fp->fdp_manager != kFdpManagerIoThread)
+                continue;
+            if(!FD_ISSET(fd, &readset))
+                continue;
+            if(fp->fdp_callback_in_audio_thread)
+            {
+                // temporarily transfer ownership
+                SYNC_STORE(&fp->fdp_manager, kFdpManagerAudioThread);
+            }
+            else
+            {
+                int fd = fp->fdp_fd;
+                t_rbskt* rbskt = fp->fdp_rbskt;
+                ring_buffer* rb = rbskt->rs_rb;
+                unsigned int size = rb_available_to_write(rb);
+                if(size > bufsize) {
+                    if(iothreadbuf_resize(size))
+                    {
+                        size = bufsize;
+                        if(rbskt->rs_preserve_boundaries)
+                            post("Incomplete packet received\n");
```

Comment:
`post` and `error` are not thread-safe! If they are not called on the audio thread, you have to lock/unlock Pd.

Reply:
good to know, thanks

Reply:
If they are just debug messages, you could also print to stderr instead. Only use `error` and `post` for things which the end user should see in the Pd console.

Reply:
I've done that for now. 755347ef5f851bc5c7070222670a5865b1a66127

Reply:
Cool!